crates/pathfinder/src/rpc.rs

```diff
@@ -67,281 +70,302 @@ impl<Context: Send + Sync + 'static> RpcModuleWrapper<Context> {
     }
 }
 
-/// Starts the HTTP-RPC server.
-pub async fn run_server(
+pub struct RpcServer {
     addr: SocketAddr,
     api: RpcApi,
-) -> Result<(HttpServerHandle, SocketAddr), anyhow::Error> {
-    let server = HttpServerBuilder::default()
-        .build(addr)
-        .await
-        .map_err(|e| match e {
-            jsonrpsee::core::Error::Transport(_) => {
-                use std::error::Error;
+    middleware: MaybeRpcMetricsMiddleware,
+}
+
+impl RpcServer {
+    pub fn new(addr: SocketAddr, api: RpcApi) -> Self {
+        Self {
+            addr,
+            api,
+            middleware: MaybeRpcMetricsMiddleware::Noop,
+        }
+    }
+
+    pub fn set_middleware(self, middleware: RpcMetricsMiddleware) -> Self {
```

Comment:
Think this more builder patterny idiomatic naming. Ah okay I see you're just copying `jsonrpsee` naming.. Hmm; anyway something to consider.
```suggestion
    pub fn with_middleware(self, middleware: RpcMetricsMiddleware) -> Self {
```

Reply:
Agreed, there's not at all discussed predecent for `with_` :)

Reply:
`with_` is more idiomatic, I did copy the `jsonrpsee` convention.